### PR TITLE
GetEI fails on noisy functions

### LINKED ISSUE #31429: Bug in getEi (v2) on noisy monitor signal
**Original reporter:** [ISIS]/[Ross Stewart]

### Expected behavior
Load the run [MAR41449.nxs](\\Olympic\Babylon5\Public\abuts\MAR41449.nxs) (should be available in local ISIS network)
and try simple script:
```
Load(Filename='MAP41449.nxs', OutputWorkspace='MAP41449', LoadMonitors=True)
GetEi(InputWorkspace='MAP41449_monitors', Monitor1Spec=41475, Monitor2Spec=41476, EnergyEstimate=70)
```
Should add ei log to the monitor workspace.
### Actual behavior
the script generates error, stating that monitor workspace contains NaN-s. No NanS are actually there

### Platforms affected
all

### PR DESCRIPTION
**Pull request by abuts**
**Description of work.**
Additional logic has been added to the GetEi2 algorithm to ensure that a divide-by-zero error doesn't occur in rare circumstances when a local maxium on the side of a peak meets cirtain criteria.

**Report to:** [Ross Stewart]

**To test:**
Run the sample script in the linked issue and ensure Ei is calculated with no errors are thrown. Unit tests have been added to recreate the bug using a toy dataset.

Fixes #31429. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/GetEi2.cpp [hunk 1]
```diff
@@ -532,6 +532,16 @@ double GetEi2::calculatePeakWidthAtHalfHeight(const API::MatrixWorkspace_sptr &d
...
       ip1--;
     }
+    // similarly, noise may cause two points around the
+    while (peak_y[ip1] == peak_y[ip2]) { // TODO: shold be smoothing with runnign average here
+      g_log.warning() << "A peak with a constant values on the trailing edge has "
+                         "been found. The estimation of the "
+                      << "half-height point will not be as accurate.\n";
+      ip1--;
+      if (ip1 < 0)
+        throw std::invalid_argument("trailing edge values are equal unil the start of the peak");
```
Comment (thomashampson) [inline Framework/Algorithms/src/GetEi2.cpp:542]: Typo in exception message.

File: Framework/Algorithms/src/GetEi2.cpp [hunk 2]
```diff
@@ -532,6 +532,16 @@ double GetEi2::calculatePeakWidthAtHalfHeight(const API::MatrixWorkspace_sptr &d
                       << "half-height point will not be as accurate.\n";
       ip1--;
     }
+    // similarly, noise may cause two points around the
+    while (peak_y[ip1] == peak_y[ip2]) { // TODO: shold be smoothing with runnign average here
```
Comment (thomashampson) [inline Framework/Algorithms/src/GetEi2.cpp:536]: Remove the TODO and either:
- Improve the implementation as part of this PR
- Add a new github issue detailing the idea you had

File: Framework/Algorithms/src/GetEi2.cpp [hunk 3]
```diff
@@ -565,6 +575,15 @@ double GetEi2::calculatePeakWidthAtHalfHeight(const API::MatrixWorkspace_sptr &d
                       << "half-height point will not be as accurate.\n";
       im1++;
     }
+    while (peak_y[im1] == peak_y[im2]) { // TODO: shoile be smoothing with runnign average here
```
Comment (thomashampson) [inline Framework/Algorithms/src/GetEi2.cpp:578]: Remove the TODO and either:
- Improve the implementation as part of this PR
- Add a new github issue detailing the idea you had

**Review by thomashampson — DISMISSED** [review]
The following release notes file will need an entry under bug fixes for algorithms:
mantid\docs\source\release\v6.2.0\framework.rst

This kind of edge case bug lends itself well to adding unit tests to check, ideally we would have tests added as part of this PR.

File: Framework/Algorithms/src/GetEi2.cpp (hunk 2 above)
Comment (abuts) [reply Framework/Algorithms/src/GetEi2.cpp:536]: This are fair comments, but unfortunately I do not have time to make such substantial changes. I have user request to fulfill and very short time to address. 

Would be nice to have unit test, bit the function is very specific

Comment (abuts) [reply Framework/Algorithms/src/GetEi2.cpp:536]: Probably one needs to take the data from the dataset and hardcode them to the unit test

**Comment by abuts** [comment]
there is the bug which is non-trivial one. It affects multiple users. The solution is not very nice but I can not spend more then a week to write proper solution.

**Comment by thomashampson** [comment]
Unassigning myself as reviewer becasue I've made code changes.

File: Framework/Algorithms/src/GetEi2.cpp (hunk 3 above)
Comment (thomashampson) [reply Framework/Algorithms/src/GetEi2.cpp]: On reflection, the "crude" changes to fix the bug are fine in the context of this algorthm. Any issue to improve the algorithm need to be directly related to this issue.

**Review by martyngigg — APPROVED** [review]
Merging as we have new tests and release notes plus something that at least works with noisier data.

**Comment by martyngigg** [comment]
@abuts If we need to investigate a new way of finding Ei then please raise this through David Voneshan so that he can discuss priorities with @SilkeSchomann.

**Pull request merged**